File: go-backend-interview-prep/08_http_and_backend/05_json_api_crud_in_memory/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Item is the resource we CRUD.
type Item struct {
	ID    int     `json:"id"`
	Name  string  `json:"name"`
	Price float64 `json:"price"`
}

// Store holds items in memory, protected by a mutex.
var (
	store  = map[int]Item{}
	mu     sync.RWMutex
	nextID = 1
)

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

// --- Handlers ---

func listItems(w http.ResponseWriter, r *http.Request) {
	mu.RLock()
	defer mu.RUnlock()
	items := make([]Item, 0, len(store))
	for _, item := range store {
		items = append(items, item)
	}
	writeJSON(w, http.StatusOK, items)
}

func getItem(w http.ResponseWriter, r *http.Request) {
	id, err := strconv.Atoi(r.PathValue("id"))
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid id"})
		return
	}
	mu.RLock()
	item, ok := store[id]
	mu.RUnlock()
	if !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "not found"})
		return
	}
	writeJSON(w, http.StatusOK, item)
}

func createItem(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Name  string  `json:"name"`
		Price float64 `json:"price"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid json"})
		return
	}
	if req.Name == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "name required"})
		return
	}

	mu.Lock()
	item := Item{ID: nextID, Name: req.Name, Price: req.Price}
	store[nextID] = item
	nextID++
	mu.Unlock()

	writeJSON(w, http.StatusCreated, item)
}

func updateItem(w http.ResponseWriter, r *http.Request) {
	id, err := strconv.Atoi(r.PathValue("id"))
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid id"})
		return
	}

	var req struct {
		Name  string  `json:"name"`
		Price float64 `json:"price"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid json"})
		return
	}

	mu.Lock()
	defer mu.Unlock()
	item, ok := store[id]
	if !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "not found"})
		return
	}
	if req.Name != "" {
		item.Name = req.Name
	}
	if req.Price != 0 {
		item.Price = req.Price
	}
	store[id] = item
	writeJSON(w, http.StatusOK, item)
}

func deleteItem(w http.ResponseWriter, r *http.Request) {
	id, err := strconv.Atoi(r.PathValue("id"))
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid id"})
		return
	}

	mu.Lock()
	defer mu.Unlock()
	if _, ok := store[id]; !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": "not found"})
		return
	}
	delete(store, id)
	writeJSON(w, http.StatusOK, map[string]string{"deleted": strconv.Itoa(id)})
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("GET /items", listItems)
	mux.HandleFunc("GET /items/{id}", getItem)
	mux.HandleFunc("POST /items", createItem)
	mux.HandleFunc("PUT /items/{id}", updateItem)
	mux.HandleFunc("DELETE /items/{id}", deleteItem)

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("CRUD API on http://%s\n", addr)
	fmt.Println("  POST   /items       -> create")
	fmt.Println("  GET    /items       -> list")
	fmt.Println("  GET    /items/{id}  -> get")
	fmt.Println("  PUT    /items/{id}  -> update")
	fmt.Println("  DELETE /items/{id}  -> delete")

	server := &http.Server{Handler: mux}

	// Self-test demo
	go func() {
		time.Sleep(100 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo ---")

		// Create
		resp, _ := http.Post(base+"/items", "application/json",
			strings.NewReader(`{"name":"widget","price":9.99}`))
		if resp != nil {
			var item Item
			json.NewDecoder(resp.Body).Decode(&item)
			resp.Body.Close()
			fmt.Printf("  CREATE: %+v (status %d)\n", item, resp.StatusCode)
		}

		// Create another
		resp2, _ := http.Post(base+"/items", "application/json",
			strings.NewReader(`{"name":"gadget","price":19.99}`))
		if resp2 != nil {
			resp2.Body.Close()
		}

		// List
		resp3, _ := http.Get(base + "/items")
		if resp3 != nil {
			var items []Item
			json.NewDecoder(resp3.Body).Decode(&items)
			resp3.Body.Close()
			fmt.Printf("  LIST:   %d items\n", len(items))
		}

		// Get
		resp4, _ := http.Get(base + "/items/1")
		if resp4 != nil {
			var item Item
			json.NewDecoder(resp4.Body).Decode(&item)
			resp4.Body.Close()
			fmt.Printf("  GET:    %+v\n", item)
		}

		// Update
		req, _ := http.NewRequest("PUT", base+"/items/1",
			strings.NewReader(`{"name":"super widget","price":14.99}`))
		req.Header.Set("Content-Type", "application/json")
		resp5, _ := http.DefaultClient.Do(req)
		if resp5 != nil {
			var item Item
			json.NewDecoder(resp5.Body).Decode(&item)
			resp5.Body.Close()
			fmt.Printf("  UPDATE: %+v\n", item)
		}

		// Delete
		delReq, _ := http.NewRequest("DELETE", base+"/items/2", nil)
		resp6, _ := http.DefaultClient.Do(delReq)
		if resp6 != nil {
			var result map[string]string
			json.NewDecoder(resp6.Body).Decode(&result)
			resp6.Body.Close()
			fmt.Printf("  DELETE: %v\n", result)
		}

		fmt.Println("\nDemo done.")
		server.Close()
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}

File: go-backend-interview-prep/08_http_and_backend/05_json_api_crud_in_memory/main.py
"""JSON API CRUD (in-memory) -- Python equivalent of the Go example."""

import json
import re
import threading
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler

# In-memory store (thread-safe with lock)
store: dict[int, dict] = {}
lock = threading.Lock()
next_id = 1


def json_response(handler, status: int, data) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


def read_json(handler) -> dict:
    length = int(handler.headers.get("Content-Length", 0))
    return json.loads(handler.rfile.read(length)) if length else {}


class CRUDHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        # GET /items -- list
        if self.path == "/items":
            with lock:
                items = list(store.values())
            json_response(self, 200, items)
            return

        # GET /items/{id}
        m = re.match(r"^/items/(\d+)$", self.path)
        if m:
            item_id = int(m.group(1))
            with lock:
                item = store.get(item_id)
            if item is None:
                json_response(self, 404, {"error": "not found"})
                return
            json_response(self, 200, item)
            return

        json_response(self, 404, {"error": "not found"})

    def do_POST(self):
        if self.path == "/items":
            global next_id
            body = read_json(self)
            name = body.get("name", "")
            if not name:
                json_response(self, 400, {"error": "name required"})
                return
            price = body.get("price", 0)
            with lock:
                item = {"id": next_id, "name": name, "price": price}
                store[next_id] = item
                next_id += 1
            json_response(self, 201, item)
            return
        json_response(self, 404, {"error": "not found"})

    def do_PUT(self):
        m = re.match(r"^/items/(\d+)$", self.path)
        if m:
            item_id = int(m.group(1))
            body = read_json(self)
            with lock:
                item = store.get(item_id)
                if item is None:
                    json_response(self, 404, {"error": "not found"})
                    return
                if body.get("name"):
                    item["name"] = body["name"]
                if body.get("price"):
                    item["price"] = body["price"]
                store[item_id] = item
            json_response(self, 200, item)
            return
        json_response(self, 404, {"error": "not found"})

    def do_DELETE(self):
        m = re.match(r"^/items/(\d+)$", self.path)
        if m:
            item_id = int(m.group(1))
            with lock:
                if item_id not in store:
                    json_response(self, 404, {"error": "not found"})
                    return
                del store[item_id]
            json_response(self, 200, {"deleted": str(item_id)})
            return
        json_response(self, 404, {"error": "not found"})

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), CRUDHandler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"CRUD API on {base}")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    print("\n--- Demo ---")

    # Create
    req = urllib.request.Request(
        f"{base}/items",
        data=json.dumps({"name": "widget", "price": 9.99}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    resp = urllib.request.urlopen(req)
    print(f"  CREATE: {resp.read().decode().strip()}")

    # Create another
    req2 = urllib.request.Request(
        f"{base}/items",
        data=json.dumps({"name": "gadget", "price": 19.99}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    urllib.request.urlopen(req2).read()

    # List
    resp3 = urllib.request.urlopen(f"{base}/items")
    items = json.loads(resp3.read())
    print(f"  LIST:   {len(items)} items")

    # Get
    resp4 = urllib.request.urlopen(f"{base}/items/1")
    print(f"  GET:    {resp4.read().decode().strip()}")

    # Update
    req5 = urllib.request.Request(
        f"{base}/items/1",
        data=json.dumps({"name": "super widget", "price": 14.99}).encode(),
        headers={"Content-Type": "application/json"},
        method="PUT",
    )
    resp5 = urllib.request.urlopen(req5)
    print(f"  UPDATE: {resp5.read().decode().strip()}")

    # Delete
    req6 = urllib.request.Request(f"{base}/items/2", method="DELETE")
    resp6 = urllib.request.urlopen(req6)
    print(f"  DELETE: {resp6.read().decode().strip()}")

    server.shutdown()
    print("Done.")


if __name__ == "__main__":
    main()
